File: Chap7-Miscellaneous/7-2PainterAppCursors/shapelist.cpp
#include "shapelist.h"
#include <QDrag>
#include <QMimeData>
#include <QDebug>

ShapeList::ShapeList(QWidget *parent) : QListWidget(parent)
{
    setMinimumWidth(200);
    setMaximumWidth(300);
    setSelectionMode(QAbstractItemView::SingleSelection);
    setViewMode(QListView::IconMode);

}

void ShapeList::startDrag(Qt::DropActions supportedActions)
{
    Q_UNUSED(supportedActions);
    QList<QListWidgetItem *> items = selectedItems();
    if(items.count()>0){

        QDrag * drag = new QDrag(this);
        QMimeData * mimeData = new QMimeData;

        QListWidgetItem * item = items[0];

        int key = item->data(Qt::UserRole).toInt();

        mimeData->setProperty("Key",QVariant::fromValue(key));

        drag->setMimeData(mimeData);
        QPixmap pix = item->icon().pixmap( 50,50 );
        drag->setPixmap( pix );

        drag->setHotSpot(pix.rect().center());

        if(drag->exec() == Qt::IgnoreAction){
            qDebug() << "Drag ignored";
        }
    }

}

File: Chap7-Miscellaneous/7-1PainterAppUiCleaned/shapelist.py
from PySide6.QtWidgets import QListWidget, QListView, QAbstractItemView
from PySide6.QtCore import Qt, QMimeData
from PySide6.QtGui import QDrag, QPixmap

class ShapeList(QListWidget):
    """List widget for selecting shapes with drag and drop capability"""
    
    def __init__(self, parent=None):
        super().__init__(parent)
        
        self.setMinimumWidth(200)
        self.setMaximumWidth(300)
        self.setSelectionMode(QAbstractItemView.SelectionMode.SingleSelection)
        self.setViewMode(QListView.ViewMode.IconMode)
    
    def startDrag(self, supported_actions):
        """Start drag and drop operation with shape data"""
        items = self.selectedItems()
        if len(items) > 0:
            drag = QDrag(self)
            mime_data = QMimeData()
            
            item = items[0]
            key = item.data(Qt.ItemDataRole.UserRole)
            
            # Store the key as property
            mime_data.setProperty("Key", key)
            
            drag.setMimeData(mime_data)
            pixmap = item.icon().pixmap(50, 50)
            drag.setPixmap(pixmap)
            
            # Set hotspot to center of the pixmap
            drag.setHotSpot(pixmap.rect().center())
            
            if drag.exec() == Qt.DropAction.IgnoreAction:
                print("Drag ignored")